### ISSUE TITLE
Incorrect error for return type with async-stream

### ISSUE THREAD

**Issue by malthe** [intellij-rust/intellij-rust#6669]
## Environment

* **IntelliJ Rust plugin version:**  0.3.138.3572-203
* **Rust toolchain version:** 1.44.0-nightly
* **IDE name and version:** CLion 2020.3
* **Operating system:** MacOS

## Problem description

IDE incorrectly reports an error in return type:
```
Function returns () instead of impl Stream<Item = u32> 
```

This seems to have something to do with the use of the [stream!](https://docs.rs/async-stream/0.1.0/async_stream/) macro.

## Steps to reproduce

```rust
use async_std::task;
use async_stream::stream;

use futures::stream::futures_unordered::FuturesUnordered;
use futures::stream::{
    Stream,
    StreamExt
};

use rand::distributions::{Distribution, Uniform};
use std::time::Duration;
use std::thread;

async fn sleep_and_print(future_number: u32, sleep_millis: u64) -> u32 {
    let sleep_duration = Duration::from_millis(sleep_millis);
    task::sleep(sleep_duration).await;
    println!("Future {} slept for {} ms on {:?}", future_number, sleep_millis, thread::current().id());
    future_number
}

pub fn test() -> impl Stream<Item = u32> {
    let mut futures = FuturesUnordered::new();

    let between = Uniform::from(500..10_000);
    let mut rng = rand::thread_rng();

    for future_number in 0..10 {
        let sleep_millis = between.sample(&mut rng);
        futures.push(sleep_and_print(future_number, sleep_millis));
    }

    stream! {
        while let Some(value) = futures.next().await {
             yield value
        }
    }
}
```

The `for` loop is what's reported as an error:

<img width="759" alt="image" src="https://user-images.githubusercontent.com/26405/104602118-f5800380-5672-11eb-891b-b69e13d87f53.png">
